dlmalloc_nonreuse.c

```diff
@@ -3219,52 +3394,71 @@ dlfree(void* mem) {
               set_size_and_clear_pdirty_of_dirty_chunk(p, psize);
               set_pdirty(next);
             }
-
-            insert_freebuf_chunk(fm, p);
-            goto postaction;
-          }
-        }
-      } else {
-        insert_freebuf_chunk(fm, p);
-        goto postaction;
+          } else
+            CORRUPTION_ERROR_ACTION(fm);
+        } else
+          USAGE_ERROR_ACTION(fm, p);
       }
-    erroraction:
-      USAGE_ERROR_ACTION(fm, p);
-    postaction:
+      insert_freebuf_chunk(fm, p);
       if(fm->freebufbytes > (size_t)(fm->footprint*DEFAULT_FREEBUF_PERCENT)) {
-#if SWEEP_STATS
-        fm->sweepTimes++;
-        fm->sweptBytes += fm->footprint;
-#endif // SWEEP_STATS
-        mchunkptr freebin = &fm->freebufbin;
-        mchunkptr thePtr = freebin->fd;
-        while(thePtr != freebin) {
-          shadow_paint(thePtr, chunksize(thePtr));
-          thePtr = thePtr->fd;
-        }
-        while(freebin->fd != freebin) {
-          mchunkptr ret = freebin->fd;
-          unlink_first_freebuf_chunk(fm, freebin, ret);
-          size_t theSize = chunksize(ret);
-          shadow_clear(ret, theSize);
-          mchunkptr theNext = chunk_plus_offset(ret, theSize);
-          if(!is_mmapped(ret)) {
-            assert(cdirty(ret));
-            assert(pdirty(theNext));
-            clear_cdirty(ret);
-            clear_pdirty(theNext);
-          }
-          // Have to do free_internal after unlinking, otherwise the circular
-          // list of freebufbin will get corrupted.
-          dlfree_internal(chunk2mem(ret));
-        }
+	malloc_revoke_internal();
       }
       POSTACTION(fm);
     }
   }
-#if !FOOTERS
-#undef fm
-#endif /* FOOTERS */
+}
+
+static void
+malloc_revoke_internal(void) {
+#if SWEEP_STATS
+  gm->sweepTimes++;
+  gm->sweptBytes += gm->footprint;
+#endif // SWEEP_STATS
+  mchunkptr freebin = &gm->freebufbin;
+  for (mchunkptr thePtr = freebin->fd; thePtr != freebin; thePtr = thePtr->fd) {
+#ifndef __CHERI_PURE_CAPABILITY__
+    shadow_paint(thePtr, chunksize(thePtr));
+#else
+#ifdef CAPREVOKE
+    vaddr_t addr = __builtin_cheri_address_get(chunk2mem(thePtr));
+    caprev_shadow_nomap_set_raw(thePtr->pad, addr, addr + chunksize(thePtr));
+#endif
+#endif
+  }
+
+  struct caprevoke_stats crst;
+  caprevoke(CAPREVOKE_LAST_PASS|CAPREVOKE_IGNORE_START, 0, &crst);
```

Comment:
This should probably, strictly, be something like the following, just in case someone else is doing revocation in the same process on a different thread as well.  Sorry the wiki doc didn't actually spell this out; got lost in all the rest of the writing.

```
struct caprevoke_stats crst;
uint64_t start_epoch;
caprevoke(CAPREVOKE_LAST_PASS|CAPREVOKE_IGNORE_START|CAPREVOKE_NO_WAIT_OK, 0, &crst);
start_epoch = crst.epoch_init;
while (!caprevoke_epoch_clears(crst->epoch_fini, start_epoch)) {
  caprevoke(CAPREVOKE_LAST_PASS, start_epoch, &crst);
}
```

It's very likely that the `caprevoke_epoch_clears` test reveals that no more calls are necessary; perhaps a `__builtin_expect` is useful.

Reply:
Hm.  In some cases that I don't understand, the looping form is actually necessary; we get some bad return codes from the vm in `caprevoke` that cause us to bail back to userland but if we just retry they eventually clear up.  That's not great, but it's just once at the process startup, it seems.